### PR TITLE
fix: improve lockfile v4 and use deno_semver to fix duplicate entry issue

### PR DESCRIPTION
**Pull request by dsherret**
Fixes the issue found here: https://github.com/denoland/deno_graph/pull/517#discussion_r1732057422

Also, makes the "specifiers" area less verbose for lockfile v4.

### CODE REVIEW

File: tests/specs/config_changes/RemovingOakThenDax.txt [hunk 1]
```diff
@@ -161,24 +161,20 @@
 {
   "version": "4",
   "specifiers": {
-    "jsr:@dsherret/dax": "jsr:@dsherret/dax@0.0.1",
-    "jsr:@dsherret/which@0.0.1": "jsr:@dsherret/which@0.0.1",
-    "jsr:@std/assert@0.210": "jsr:@std/assert@0.210.0",
-    "jsr:@std/assert@^0.210.0": "jsr:@std/assert@0.210.0",
```
Comment (dsherret) [inline tests/specs/config_changes/RemovingOakThenDax.txt:167]: You'll notice here that we got smarter and can tell these are actually the same version requirement now.

File: src/lib.rs [hunk 2]
```diff
@@ -75,27 +76,24 @@ pub struct NpmPackageInfo {
...
+#[derive(Debug, Clone)]
 pub struct JsrPackageInfo {
   pub integrity: String,
   /// List of package requirements found in the dependency.
   ///
   /// This is used to tell when a package can be removed from the lockfile.
-  #[serde(skip_serializing_if = "BTreeSet::is_empty")]
-  #[serde(default)]
-  pub dependencies: BTreeSet<String>,
+  pub dependencies: HashSet<JsrDepPackageReq>,
```
Comment (bartlomieju) [inline src/lib.rs:85]: No longer need to keep ordering?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: src/lib.rs (hunk 2 above)
Comment (dsherret) [reply src/lib.rs:85]: It's handled in the printer only now.

**Comment by dsherret** [comment]
I'm going to merge this one, but will follow up with a new PR that normalizes the versions.

**Pull request merged**